xarray/core/resample.py

```diff
@@ -80,12 +106,17 @@ def _upsample(self, method, *args, **kwargs):
                 self._obj = self._obj.drop_vars(k)
 
         if method == "asfreq":
-            return self.mean(self._dim)
+            return self.mean(self._dim)  # type: ignore[attr-defined]
 
         elif method in ["pad", "ffill", "backfill", "bfill", "nearest"]:
             kwargs = kwargs.copy()
-            kwargs.update(**{self._dim: upsampled_index})
-            return self._obj.reindex(method=method, *args, **kwargs)
+            if not isinstance(self._dim, str):
+                raise NotImplementedError(
+                    "Only str dimensions supported by now. Please raise an issue on Github."
+                )
+            kwargs[self._dim] = upsampled_index
+            kwargs["method"] = method
+            return self._obj.reindex(*args, **kwargs)
```

Comment:
If we remove `*args` above, this can just be:
```suggestion
            kwargs["method"] = method
            return self._obj.reindex({self._dim: upsampled_index}, **kwargs)
```

Which I believe should pass type checking, even for non-string dimension names.

For methods like `reindex` with a signature like `reindex(indexers=None, method=None, tolerance=None, copy=True, fill_value=<NA>, **indexers_kwargs)`, we generally try to avoid using `**indexers_kwargs` internally inside Xarray in favor of `indexers`, for exactly this sort of reason.

Reply:
I am now getting rid of the upsample method all together, I found that way cleaner.
If you are ok with this change, feel free to resolve this conversation, otherwise I can revert this commit.